src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,187 @@
+package duke;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+/**
+ * This class encapsulates the mechanism to parse user commands.
+ *
+ * @author Kleon Ang
+ */
+public class Parser {
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-d H:mm");
+    private final TaskList tasks;
+    private boolean toRewriteData;
+    private boolean isExit;
+    private boolean isFind;
+
+    /**
+     * Constructor for a Parser class.
+     *
+     * @param tasks A TaskList for the Parser to refer to.
+     */
+    public Parser(TaskList tasks) {
+        this.tasks = tasks;
+        this.toRewriteData = false;
+        this.isExit = false;
+        this.isFind = false;
+    }
+
+    private void list(TaskList tasks) throws DukeException {
+        if (tasks.size() == 0) {
+            throw new DukeException("There are currently no tasks in your list.");
+        }
+        String matching = this.isFind ? "matching " : "";
+        StringBuilder tasksBuilder = new StringBuilder();
+        tasksBuilder.append("Here are the ").append(matching).append("tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); ++i) {
+            String counter = String.valueOf(i + 1);
+            Task currentTask = tasks.get(i);
+            tasksBuilder.append(counter).append(".").append(currentTask);
+            if (i < tasks.size() - 1) {
+                // Append newline for all tasks before last task
+                tasksBuilder.append("\n");
+            }
+        }
+        Ui.printReply(tasksBuilder.toString());
+    }
+
+    private void done(int counter) throws DukeException {
+        if (counter <= 0 || counter > tasks.size()) {
+            throw new DukeException("Sorry, no such task of index " + counter + ".");
+        }
+        Task doneTask = tasks.get(counter - 1);
+        doneTask.markAsDone();
+        Ui.printReply("Nice! I've marked this task as done:\n  " + doneTask);
+    }
+
+    /**
+     * Getter for the rewrite status of the last parsed command.
+     *
+     * @return True if data in Storage needs to be rewritten.
+     */
+    public boolean toRewrite() {
+        return this.toRewriteData;
+    }
+
+    /**
+     * Getter for the exit status of the last parsed command.
+     *
+     * @return True if ready to exit Duke after "bye" command.
+     */
+    public boolean toExit() {
+        return this.isExit;
+    }
+
```

Comment:
I think for methods, maybe try to keep their names as verbs, such as needsRewrite and needsToExit.

Reply:
Agreed, since its a getter could also consider more default naming convention like getExitStatus(), getRewriteStatus()